### PR TITLE
feat: ai message summary

### PR DESCRIPTION
**Pull request by SebastianKrupinski**
Fix https://github.com/nextcloud/mail/issues/10387
Fix https://github.com/nextcloud/mail/issues/10388

Summarize messages using Ai

### CODE REVIEW

**Comment by GVodyanov** [comment]
image

Hovering over the sparkle or the summary text brings up the disclaimer

File: lib/Service/AiIntegrations/AiIntegrationsService.php [hunk 1]
```diff
@@ -61,6 +66,74 @@ public function __construct(ContainerInterface $container, Cache $cache, IMAPCli
...
+			$manager = $this->container->get(TaskProcessingManager::class);
+		} catch (\Throwable $e) {
+			throw new ServiceException('Task processing is not available', 0, $e);
+		}
+		try {
+			$manager->getPreferredProvider(TextToTextSummary::ID);
+		} catch (TaskProcessingException $e) {
+			throw new ServiceException('No text summary provider available');
+		}
+		if (!$client) {
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:92]: I don't see an actual usage where a client is passed. Does it exist?

**Review by ChristophWurst — COMMENTED** [review]
Quick first review. Architecture looks sound, there are a few issues with code quality.

File: lib/Listener/TaskProcessingListener.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+use OCP\TaskProcessing\TaskTypes\TextToTextSummary;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @template-implements IEventListener<Event|NewMessagesSummarizeListener>
+ */
+class TaskProcessingListener implements IEventListener {
+
+	public function __construct(
+		protected LoggerInterface $logger,
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:26]: Done

File: lib/Listener/TaskProcessingListener.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		protected MessageMapper $messageStore,
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:27]: Done

File: lib/Listener/TaskProcessingListener.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+		if ($task->getAppId() !== Application::APP_ID) {
+			return;
+		}
+
+		if ($task->getTaskTypeId() !== TextToTextSummary::ID) {
+			return;
+		}
+
+		list($type, $id) = explode(':', $task->getCustomId());
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:46]: Done

File: lib/Listener/TaskProcessingListener.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		$userId = $task->getUserId();
+		$summary = $task->getOutput()['output'];
+
+		match ($type) {
+			'message' => $this->handleMessageSummary($userId, (int)$id, $summary),
+		};
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:52]: Done

File: lib/Listener/TaskProcessingListener.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	}
+
+	protected function handleMessageSummary(string $userId, int $id, string $summary) {
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:56]: Done

File: lib/Listener/TaskProcessingListener.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		$messages = $this->messageStore->findByIds($userId, [$id], '');
+
+		if (count($messages) !== 1) {
+			return;
+		}
+
+		$message = $messages[0];
+		$message->setSummary($summary);
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:64]: Done

File: lib/Service/AiIntegrations/AiIntegrationsService.php [hunk 8]
```diff
@@ -61,6 +66,74 @@ public function __construct(ContainerInterface $container, Cache $cache, IMAPCli
...
+	 * @param array<Message> $messages
+	 * @param Horde_Imap_Client_Socket $client
+	 *
+	 * @return null|string
+	 *
+	 * @throws ServiceException
+	 */
+	public function summarizeMessages(Account $account, array $messages, Horde_Imap_Client_Socket $client = null): void {
+		try {
+			$manager = $this->container->get(TaskProcessingManager::class);
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:83]: Done

File: lib/Service/AiIntegrations/AiIntegrationsService.php [hunk 9]
```diff
@@ -61,6 +66,74 @@ public function __construct(ContainerInterface $container, Cache $cache, IMAPCli
...
+		} catch (\Throwable $e) {
+			throw new ServiceException('Task processing is not available', 0, $e);
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:85]: Done

File: lib/Service/AiIntegrations/AiIntegrationsService.php [hunk 10]
```diff
@@ -61,6 +66,74 @@ public function __construct(ContainerInterface $container, Cache $cache, IMAPCli
...
+		}
+		try {
+			$manager->getPreferredProvider(TextToTextSummary::ID);
+		} catch (TaskProcessingException $e) {
+			throw new ServiceException('No text summary provider available');
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:90]: Done

File: lib/Service/AiIntegrations/AiIntegrationsService.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:92]: The usage does not exist, I was just trying to future proof it. I'll remove it.

File: lib/Service/AiIntegrations/AiIntegrationsService.php [hunk 11]
```diff
@@ -61,6 +66,74 @@ public function __construct(ContainerInterface $container, Cache $cache, IMAPCli
...
+		}
+		if (!$client) {
+			$client = $this->clientFactory->getClient($account);
+		}
+		try {
+			foreach ($messages as $entry) {
+
+				if (!empty($entry->getSummary())) {
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:98]: Done

File: lib/Service/AiIntegrations/AiIntegrationsService.php [hunk 12]
```diff
@@ -61,6 +66,74 @@ public function __construct(ContainerInterface $container, Cache $cache, IMAPCli
...
+						'input' => $prompt,
+					],
+					Application::APP_ID,
+					$userId,
+					'message:' . (string)$messageLocalId
+				);
+				$manager->scheduleTask($task);
+			}
+		} finally {
+			$client->logout();
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply lib/Service/AiIntegrations/AiIntegrationsService.php:133]: True, irrelevant now that function has been removed

File: appinfo/info.xml [hunk 13]
```diff
@@ -34,7 +34,7 @@ The rating depends on the installed text processing backend. See [the rating ove
 
 Learn more about the Nextcloud Ethical AI Rating [in our blog](https://nextcloud.com/blog/nextcloud-ethical-ai-rating/).
 	]]></description>
-	<version>4.2.0-alpha.0</version>
+	<version>4.2.1-alpha.0</version>
```
Comment (ChristophWurst) [inline appinfo/info.xml:37]:
```suggestion
	<version>4.2.0-alpha.1</version>
```

main will become 4.2.0. The suffix is for internal versioning during the development phase. 4.2.1 will be the first patch update of 4.2.0

File: lib/Listener/TaskProcessingListener.php [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		if ($task->getTaskTypeId() !== TextToTextSummary::ID) {
+			return;
+		}
+
+		if ($task->getCustomId() !== null) {
+			[$type, $id] = explode(':', $task->getCustomId());
+		} else {
+			$this->logger->info('Error handling task processing event custom id missing');
```
Comment (ChristophWurst) [inline lib/Listener/TaskProcessingListener.php:50]: For debuggability I suggest to log the value of the custom ID
```suggestion
			$this->logger->info('Error handling task processing event custom id missing', ['taskCustomId' => $task->getCustomid(), ]);
```

File: lib/Listener/TaskProcessingListener.php [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+			return;
+		}
+		if ($type === null || $id === null) {
+			$this->logger->info('Error handling task processing event custom id is invalid');
```
Comment (ChristophWurst) [inline lib/Listener/TaskProcessingListener.php:54]: same

File: lib/Listener/TaskProcessingListener.php [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+		if ($task->getUserId() !== null) {
+			$userId = $task->getUserId();
+		} else {
+			$this->logger->info('Error handling task processing event user id missing');
+			return;
+		}
+		if ($task->getOutput() !== null) {
+			$output = $task->getOutput();
+			if (isset($task->getOutput()['output'])) {
+				/* @psalm-suppress PossiblyInvalidCast */
```
Comment (ChristophWurst) [inline lib/Listener/TaskProcessingListener.php:66]: Can you elaborate on the suppressed check?

File: appinfo/info.xml (hunk 13 above)
Comment (SebastianKrupinski) [reply appinfo/info.xml:37]: Done

File: lib/Listener/TaskProcessingListener.php (hunk 14 above)
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:50]: Done

File: lib/Listener/TaskProcessingListener.php (hunk 15 above)
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:54]: Done

File: lib/Listener/TaskProcessingListener.php (hunk 16 above)
Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php:66]: This was my last resort to satisfy psalm and it's still not happy.

image

Comment (ChristophWurst) [reply lib/Listener/TaskProcessingListener.php]: I stand with Psalm. https://github.com/nextcloud/server/blob/db683da757d854ff0e86d78a8b7053e2677731ac/lib/public/TextProcessing/Task.php#L116-L118 the method returns null or a string. You are interpreting the string as an array. How is that possible? The PHP return type hint won't let `getOutput` return anything but null or string. Other values cause a type error.

Comment (SebastianKrupinski) [reply lib/Listener/TaskProcessingListener.php]: That's the wrong class... This is the correct one... https://github.com/nextcloud/server/blob/db683da757d854ff0e86d78a8b7053e2677731ac/lib/public/TaskProcessing/Task.php#L155

Anyway, its fixed, psalm is happy

File: appinfo/info.xml [hunk 17]
```diff
@@ -34,7 +34,7 @@ The rating depends on the installed text processing backend. See [the rating ove
 
 Learn more about the Nextcloud Ethical AI Rating [in our blog](https://nextcloud.com/blog/nextcloud-ethical-ai-rating/).
 	]]></description>
-	<version>4.2.0-alpha.0</version>
+	<version>4.2.0-alpha.1</version>
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply appinfo/info.xml:37]: Done

**Review by ChristophWurst — APPROVED** [review]

**Comment by hamza221** [comment]
This is not checking if the admin config for llm is enabled

**Pull request merged**